packages/flutter/lib/src/cupertino/date_picker.dart

```diff
@@ -497,69 +498,83 @@ class CupertinoDatePicker extends StatefulWidget {
         }
       case _PickerColumnType.dayPeriod:
-        longestText =
+       final String longestText =
           localizations.anteMeridiemAbbreviation.length > localizations.postMeridiemAbbreviation.length
             ? localizations.anteMeridiemAbbreviation
             : localizations.postMeridiemAbbreviation;
+        longestTexts.add(longestText);
       case _PickerColumnType.dayOfMonth:
         int longestDayOfMonth = 1;
         for (int i = 1; i <=31; i++) {
           final String dayOfMonth = localizations.datePickerDayOfMonth(i);
-          if (longestText.length < dayOfMonth.length) {
-            longestText = dayOfMonth;
-            longestDayOfMonth = i;
-          }
+          longestTexts.add(dayOfMonth);
+          longestDayOfMonth = dayOfMonth.length > longestDayOfMonth ? dayOfMonth.length : longestDayOfMonth;
         }
         if (showDayOfWeek) {
           for (int wd = 1; wd < DateTime.daysPerWeek; wd++) {
             final String dayOfMonth = localizations.datePickerDayOfMonth(longestDayOfMonth, wd);
-            if (longestText.length < dayOfMonth.length) {
-              longestText = dayOfMonth;
-            }
+            longestTexts.add(dayOfMonth);
           }
         }
       case _PickerColumnType.month:
         for (int i = 1; i <= 12; i++) {
           final String month = standaloneMonth
               ? localizations.datePickerStandaloneMonth(i)
               : localizations.datePickerMonth(i);
-          if (longestText.length < month.length) {
-            longestText = month;
-          }
+          longestTexts.add(month);
         }
       case _PickerColumnType.year:
-        longestText = localizations.datePickerYear(2018);
+        final String longestText = localizations.datePickerYear(2018);
+        longestTexts.add(longestText);
+    }
+
+    assert(longestTexts.isNotEmpty, 'longestTexts should not be empty');
+
+    for (final String text in _getLongestTexts(longestTexts)) {
+      final double width = _computeTextWidth(text: text, context: context);
+      longestTextWidth = math.max(longestTextWidth, width);
     }
 
-    assert(longestText != '', 'column type is not appropriate');
+    return longestTextWidth;
+  }
 
+  /// Compute the width of the text.
+  ///
+  /// This method is used to compute the width of the text in the picker.
+  /// It is used to estimate the minimum width that each column needs to layout
+  /// its content.
+  static double _computeTextWidth({required String text, required BuildContext context}) {
     return TextPainter.computeMaxIntrinsicWidth(
       text: TextSpan(
+        text: text,
         style: _themeTextStyle(context),
-        text: longestText,
       ),
       textDirection: Directionality.of(context),
     );
   }
+
+  // Get maximum three texts from the date picker texts.
+  // This is used to get the longest words from the date picker texts.
+  static List<String> _getLongestTexts(List<String> longestTexts) {
```

Comment:
I'm no linguist, but this doesn't seem right:

1. Some characters in the localized text may have a length of 2 (instead of one) in UTF16: https://en.wikibooks.org/wiki/Unicode/Character_reference/10000-10FFF
2. why 3 specifically?